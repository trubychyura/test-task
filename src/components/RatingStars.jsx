import { useState } from 'react';
import { Form } from 'react-bootstrap';
import { Field } from 'formik';

import {
  StyledContainer,
  StyledRadio,
  StyledStar,
} from '../styled/RatingStars';

const RatingStars = ({ stars, name }) => {
  const [hover, setHover] = useState(null);

  if (stars !== undefined) {
    return (
      <StyledContainer>
        {[...Array(5)].map((_, i) => (
          <StyledStar
            value={i + 1}
            rating={Math.round(stars)}
            hover={null}
            key={i}
          />
        ))}
      </StyledContainer>
    );
  }

  return (
    <StyledContainer>
      <Field name={name}>
        {({ field }) =>
          [...Array(5)].map((_, i) => {
            let rating = i + 1;
            return (
              <Form.Label htmlFor={`formRadio${i}`} key={i}>
                <StyledStar
                  value={rating}
                  rating={field.value}
                  hover={hover}
                  onMouseEnter={() => setHover(rating)}
                  onMouseLeave={() => setHover(0)}
                />
                <StyledRadio
                  type='radio'
                  id={`formRadio${i}`}
                  {...field}
                  value={rating}
                  checked={rating === field.value}
                />
              </Form.Label>
            );
          })
        }
      </Field>
    </StyledContainer>
  );
};

export default RatingStars;
