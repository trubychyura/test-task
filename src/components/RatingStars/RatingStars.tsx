import { FC, useState } from 'react';
import { Field, FieldProps } from 'formik';
import { Form } from 'react-bootstrap';

import { RatingProps } from '../../types';
import {
  StyledContainer,
  StyledRadio,
  StyledStar,
} from '../../styled/StyledRating';

const RatingStars: FC<RatingProps> = ({ name, stars }) => {
  const [hover, setHover] = useState<number | null>(null);

  if (stars) {
    return (
      <StyledContainer>
        {[...Array(5)].map((_, i) => (
          <StyledStar value={i + 1} rating={Math.round(stars || 0)} key={i} />
        ))}
      </StyledContainer>
    );
  }

  return (
    <StyledContainer>
      <Field name={name}>
        {({ field }: FieldProps) =>
          [...Array(5)].map((_, i) => {
            let rating: number = i + 1;

            return (
              <Form.Label htmlFor={`formRadio${i}`} key={i}>
                <StyledStar
                  value={rating}
                  rating={field.value}
                  hover={hover}
                  onMouseEnter={() => setHover(rating)}
                  onMouseLeave={() => setHover(0)}
                />
                <StyledRadio
                  type='radio'
                  id={`formRadio${i}`}
                  {...field}
                  value={rating}
                  checked={rating === field.value}
                  data-testid={`radio${i}`}
                />
              </Form.Label>
            );
          })
        }
      </Field>
    </StyledContainer>
  );
};

export default RatingStars;
