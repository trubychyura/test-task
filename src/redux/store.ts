import { applyMiddleware, compose, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';

import reducer from './reducer';
import { watcherLoadData } from './ducks/products';

const composeEnhancers =
  (typeof window !== 'undefined' &&
    (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;

const sagaMiddleware = createSagaMiddleware();

export default createStore(
  reducer,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
);

sagaMiddleware.run(watcherLoadData);
