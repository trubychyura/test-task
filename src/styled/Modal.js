import styled from 'styled-components';
import { Col, Form, Image, Modal, Row } from 'react-bootstrap';

export const StyledModal = styled(Modal)`
  @media (max-width: 560px) {
    padding: 0   !important;
  }

  .modal-title {
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
  }

  .modal-dialog {
    max-width: 50%;
    @media (max-width: 1000px) {
      max-width: 70%;
    }
    @media (max-width: 580px) {
      max-width: 90%;
      margin: auto;
    }
  }
`;

export const StyledImage = styled(Image)`
  display: block;
  max-height: 300px;
  max-width: 100%;
  margin: auto;
`;

export const StyledForm = styled(Form)`
  width: 100%;
`;

export const StyledDescription = styled(Col)`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const StyledComments = styled(Col)`
  padding: 15px;
`;

export const StyledComentHeader = styled(Row)`
  margin: 0;
  justify-content: space-between;
`;
