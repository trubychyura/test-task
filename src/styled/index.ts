export const theme = {
  color: {
    gold: 'gold',
    silver: 'silver',
  },
  media: {
    md: '580px',
    lg: '1000px',
  },
};
