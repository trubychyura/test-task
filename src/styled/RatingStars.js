import styled from 'styled-components';
import { Star } from '@styled-icons/boxicons-solid/Star';
import { Form } from 'react-bootstrap';

export const StyledStar = styled(Star)`
  cursor: pointer;
  transition: color 200ms;

  ${({ value, rating, hover }) =>
    value <= (hover || rating) ? `color:  gold;` : `color: silver;`}
`;

export const StyledRadio = styled(Form.Check)`
  display: none;
`;

export const StyledContainer = styled(Form.Group)`
  display: flex;
`;
