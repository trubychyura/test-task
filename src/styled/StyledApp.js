import styled from 'styled-components';
import { Container, Table } from 'react-bootstrap';

export const StyledContainer = styled(Container)`
  padding: 20px 10px;
`;

export const StyledTable = styled(Table)`
  text-align: center;
  th {
    font-size: 40px;
  }

  tbody tr {
    cursor: pointer;
  }
`;
