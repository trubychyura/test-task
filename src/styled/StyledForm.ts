import styled from 'styled-components';
import { Alert, Button, Form } from 'react-bootstrap';

export const StyledForm = styled(Form)`
  width: 100%;
`;

export const StyledAlert = styled(Alert)`
  font-size: 13px;
  margin: 5px 0;
  padding: 5px;
`;

export const StyledButton = styled(Button)`
  padding-left: 30px;
  padding-right: 30px;
`;
