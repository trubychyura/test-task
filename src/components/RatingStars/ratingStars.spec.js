import { shallow, mount, render } from 'enzyme';
// import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import store from '../../redux/store';
import { theme } from '../../styled';
import RatingStars from './RatingStars';

const setUp = (props) => shallow(<RatingStars {...props} />);

describe('RatingStars component', () => {
  it('should render component with rating', () => {
    const component = setUp();
    expect(component).toMatchSnapshot();
  });

  it('should render component without rating and with name', () => {
    const component = setUp({ stars: 5 });
    expect(component).toMatchSnapshot();
  });
});
