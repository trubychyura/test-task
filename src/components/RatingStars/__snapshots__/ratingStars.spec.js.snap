// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingStars component should render component with rating 1`] = `
<styled.div>
  <Field>
    <Component />
  </Field>
</styled.div>
`;

exports[`RatingStars component should render component without rating and with name 1`] = `
<styled.div>
  <Styled(Star)
    key="0"
    rating={5}
    value={1}
  />
  <Styled(Star)
    key="1"
    rating={5}
    value={2}
  />
  <Styled(Star)
    key="2"
    rating={5}
    value={3}
  />
  <Styled(Star)
    key="3"
    rating={5}
    value={4}
  />
  <Styled(Star)
    key="4"
    rating={5}
    value={5}
  />
</styled.div>
`;
